### PR TITLE
Added memory tracker support to Windows.

### PR DESCRIPTION
**Pull request by AWoloszyn**
This greatly improves the performance of coherent memory
tracking in Windows.

### CODE REVIEW

File: core/memory_tracker/cc/memory_tracker.cpp [hunk 1]
```diff
@@ -88,7 +87,7 @@ bool MemoryTracker::RemoveTrackingRangeImpl(void* start, size_t size) {
   size_t size_aligned = GetAlignedSize(start, size, page_size_);
   dirty_pages_.RecollectIfPossible(size_aligned / page_size_);
   // TODO(qining): Add Windows support
```
Comment (ben-clayton) [inline core/memory_tracker/cc/memory_tracker.cpp]: Remove TODO

File: core/memory_tracker/cc/memory_tracker.cpp [hunk 2]
```diff
@@ -74,8 +73,8 @@ bool MemoryTracker::AddTrackingRangeImpl(void* start, size_t size) {
   dirty_pages_.Reserve(size_aligned / page_size_);
   ranges_[reinterpret_cast<uintptr_t>(start)] = size;
   // TODO(qining): Add Windows support
```
Comment (Qining) [inline core/memory_tracker/cc/memory_tracker.cpp]: We can drop this TODO?

File: core/memory_tracker/cc/memory_tracker.cpp (hunk 1 above)
Comment (Qining) [inline core/memory_tracker/cc/memory_tracker.cpp]: Drop this?

File: core/memory_tracker/cc/memory_tracker.h [hunk 3]
```diff
@@ -38,6 +40,83 @@
 namespace gapii {
 namespace TrackMemory {
 
+enum class page_protections {
+  none = 0,
+  read = 1,
+  write = 2,
+  read_write = 1 | 2
```
Comment (Qining) [inline core/memory_tracker/cc/memory_tracker.h]: I think it is still better to have them 0x0, 0x1, 0x2, 0x3?
And do we impose Google C++ style here?

File: core/memory_tracker/cc/memory_tracker.h [hunk 4]
```diff
@@ -38,6 +40,83 @@
...
+inline bool set_protection(void* p, size_t size, page_protections prot) {
+  uint32_t protections = (prot == page_protections::read) ? PROT_READ :
+                    (prot == page_protections::write) ? PROT_WRITE :
+                (prot == page_protections::read_write) ? PROT_READ | PROT_WRITE: 0;
+  return mprotect(p, size, protections); 
+}
+
+// SignalBlocker blocks the specified signal when it is constructed, and
+// unblock the signal when it is destroyed.
+// TODO(qining): Add Windows support
```
Comment (Qining) [inline core/memory_tracker/cc/memory_tracker.h]: Drop this TODO?

File: core/memory_tracker/cc/memory_tracker.h [hunk 5]
```diff
@@ -320,9 +376,13 @@ class MemoryTracker {
   // By default |track_read| is set to false.
   MemoryTracker(bool track_read = false)
       : track_read_(track_read),
+#if IS_POSIX
```
Comment (Qining) [inline core/memory_tracker/cc/memory_tracker.h]: Should we use `if (TARGET_OS == GAPID_OS_LINUX) || (TARGET_OS == GAPID_OS_ANDROID)` to keep the thing consistent?

File: core/memory_tracker/cc/memory_tracker.h [hunk 6]
```diff
@@ -16,16 +16,18 @@
 
 #include "core/cc/target.h"
 
-#if (TARGET_OS == GAPID_OS_LINUX) || (TARGET_OS == GAPID_OS_ANDROID)
 #define COHERENT_TRACKING_ENABLED
 
 #ifndef GAPII_MEMORY_TRACKER_H
 #define GAPII_MEMORY_TRACKER_H
 
+#if (TARGET_OS == GAPID_OS_LINUX) || (TARGET_OS == GAPID_OS_ANDROID)
```
Comment (ben-clayton) [inline core/memory_tracker/cc/memory_tracker.h]: Consider splitting into separate files under `core/memory_tracker/cc/posix`, `core/memory_tracker/cc/windows` like we do in other directories.

File: core/memory_tracker/cc/memory_tracker.h [hunk 7]
```diff
@@ -437,26 +497,32 @@ class MemoryTracker {
     return r;
   }
 
-  // We need a static pointer as the signal handler function passed to
-  // sigaction() must be a static function..
+  // We need a static pointer as the handler function passed to
+  // must be a static function..
   static MemoryTracker* unique_tracker;
 
+#if IS_POSIX
```
Comment (Qining) [inline core/memory_tracker/cc/memory_tracker.h]: Same as before.

File: core/memory_tracker/cc/memory_tracker.cpp (hunk 2 above)
Comment (AWoloszyn) [inline core/memory_tracker/cc/memory_tracker.cpp]: Done

File: core/memory_tracker/cc/memory_tracker.h (hunk 5 above)
Comment (AWoloszyn) [inline core/memory_tracker/cc/memory_tracker.h]: I much prefer IS_POSIX since that is what we are really testing for.
Since IS_POSIX implies TARGET_OS == GAPID_OS_LINUX) || (TARGET_OS == GAPID_OS_ANDROID

File: core/memory_tracker/cc/memory_tracker.h [hunk 8]
```diff
@@ -15,19 +15,26 @@
...
 
 #include "core/cc/target.h"
+#include "core/memory_tracker/cc/memory_protections.h"
 
 #define COHERENT_TRACKING_ENABLED
-
-#ifndef GAPII_MEMORY_TRACKER_H
-#define GAPII_MEMORY_TRACKER_H
-
+#ifdef COHERENT_TRACKING_ENABLED
```
Comment (ben-clayton) [inline core/memory_tracker/cc/memory_tracker.h]: Nit: I prefer to use:

```c++
#define COHERENT_TRACKING_ENABLED 1
#if COHERENT_TRACKING_ENABLED
```

As I believe you can configure the compile to warn if the macro doesn't exist.

File: core/memory_tracker/cc/memory_tracker.h [hunk 9]
```diff
@@ -15,19 +15,26 @@
  */
 
 #include "core/cc/target.h"
+#include "core/memory_tracker/cc/memory_protections.h"
 
 #define COHERENT_TRACKING_ENABLED
```
Comment (ben-clayton) [inline core/memory_tracker/cc/memory_tracker.h]: You want this here?

File: core/memory_tracker/cc/windows/memory_tracker.h [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+#ifndef GAPII_MEMORY_TRACKER_WINDOWS_H
+#define GAPII_MEMORY_TRACKER_WINDOWS_H
+
+#include <Windows.h>
+#include <cstdint>
+#include <atomic>
+
+namespace gapii {
+namespace TrackMemory {
```
Comment (ben-clayton) [inline core/memory_tracker/cc/windows/memory_tracker.h]: nit: namespaces are usually lowercase in our codebase.

File: core/memory_tracker/cc/memory_tracker.h [hunk 11]
```diff
@@ -311,47 +297,48 @@ class DirtyPageTable {
       current_;  // The space for the last page address stored.
 };
 
-// MemoryTracker utilizes Segfault signal on Linux to track accesses to
+// MemoryTrackerImpl utilizes Segfault signal on Linux to track accesses to
 // memories.
-class MemoryTracker {
+template<typename SpecificMemoryTracker>
```
Comment (ben-clayton) [inline core/memory_tracker/cc/memory_tracker.h]: Nit: Not sure if there's any google policies on template variable naming, but I've always used upper-case names, usually single letters, documented in the class.

Comment (AWoloszyn) [inline core/memory_tracker/cc/memory_tracker.h]: The style guide suggests naming them the same as Classes.
https://google.github.io/styleguide/cppguide.html#Type_Names

The names of all types — classes, structs, type aliases, enums, and type template parameters — have the same naming convention

**Comment by Qining** [comment]
LGTM except the enum name: `page_protections`, maybe: `PageProtections`?

**Comment by AWoloszyn** [comment]
Fixed

**Comment by ben-clayton** [comment]
Various Kokoro build failures. PTAL.

File: core/memory_tracker/cc/memory_tracker.h [hunk 12]
```diff
@@ -16,16 +16,24 @@
...
-#define COHERENT_TRACKING_ENABLED
-
 #ifndef GAPII_MEMORY_TRACKER_H
 #define GAPII_MEMORY_TRACKER_H
 
-#include <signal.h>
-#include <stdlib.h>
-#include <sys/mman.h>
-#include <unistd.h>
+#define COHERENT_TRACKING_ENABLED 1
```
Comment (ben-clayton) [inline core/memory_tracker/cc/memory_tracker.h]: nit: Indent nested pre-processor logic after `#`, it's really hard to follow without indents.
